File: bitcache/memory.py
"""Agent memory with prioritization, aging, and eviction.

Memories have importance scores that decay over time. Frequently
accessed memories get reinforced. Low-value memories are evicted
when capacity is reached.

Designed for long-running AI agents that need human-like memory:
- Remember important things
- Forget useless things
- Strengthen frequently used memories
"""

import time
from typing import Any, Dict, List, Optional

import numpy as np

from bitcache.streaming import StreamingIndex


class AgentMemory:
    """Prioritized agent memory with decay and eviction.

    Each memory has:
    - vector: embedding for similarity search
    - content: the actual text/data
    - importance: score from 0.0 to 1.0
    - access_count: how many times retrieved
    - last_accessed: timestamp of last retrieval
    - created_at: insertion timestamp

    Importance decays over time. Retrieval reinforces importance.
    When capacity is reached, lowest-priority memories are evicted.

    Args:
        dim: Vector dimensionality.
        capacity: Max memories before eviction triggers.
        decay_rate: Importance decay per day (0.0 = no decay, 1.0 = full decay).
        reinforce_amount: How much retrieval boosts importance.
    """

    def __init__(
        self,
        dim: int,
        capacity: int = 10000,
        decay_rate: float = 0.05,
        reinforce_amount: float = 0.1,
        rerank_factor: int = 10,
    ):
        self.dim = dim
        self.capacity = capacity
        self.decay_rate = decay_rate
        self.reinforce_amount = reinforce_amount

        self._index = StreamingIndex(dim=dim, rerank_factor=rerank_factor)
        self._memory_state: Dict[str, Dict[str, Any]] = {}

    def save_memory(
        self,
        vector: np.ndarray,
        content: str,
        importance: float = 0.5,
        id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a new memory.

        Args:
            vector: Embedding vector.
            content: The text content of the memory.
            importance: Initial importance score (0.0 to 1.0).
            id: Optional ID. Auto-generated if not provided.
            metadata: Optional additional metadata.

        Returns:
            Memory ID.
        """
        importance = max(0.0, min(1.0, importance))
        now = time.time()

        meta = metadata or {}
        meta["content"] = content
        meta["importance"] = importance

        mid = self._index.insert(vector, id=id, metadata=meta)

        self._memory_state[mid] = {
            "importance": importance,
            "access_count": 0,
            "last_accessed": now,
            "created_at": now,
        }

        # Evict if over capacity
        if len(self._index) > self.capacity:
            self._evict()

        return mid

    def retrieve_memory(
        self,
        query: np.ndarray,
        k: int = 5,
        min_importance: float = 0.0,
    ) -> List[Dict[str, Any]]:
        """Retrieve relevant memories and reinforce them.

        Args:
            query: Query vector.
            k: Number of memories to return.
            min_importance: Minimum importance threshold.

        Returns:
            List of memory dicts with content, importance, score.
        """
        # Apply decay before searching
        self._apply_decay()

        # Build filter if min_importance > 0
        filter_dict = None
        # Can't filter by importance directly in metadata (it's a float comparison)
        # So we fetch more and filter post-hoc

        scores, ids, metas = self._index.search(query, k=k * 3)

        results = []
        for score, mid, meta in zip(scores, ids, metas):
            if mid not in self._memory_state:
                continue

            state = self._memory_state[mid]
            current_importance = state["importance"]

            if current_importance < min_importance:
                continue

            # Reinforce this memory (it was useful)
            self._reinforce(mid)

            results.append({
                "id": mid,
                "content": meta.get("content", ""),
                "importance": current_importance,
                "score": float(score),
                "access_count": state["access_count"],
                "metadata": {k: v for k, v in meta.items() if k not in ("content", "importance")},
            })

            if len(results) >= k:
                break

        return results

    def reinforce_memory(self, id: str, amount: Optional[float] = None) -> bool:
        """Manually reinforce a memory's importance.

        Args:
            id: Memory ID.
            amount: Reinforcement amount. Uses default if None.

        Returns:
            True if memory exists and was reinforced.
        """
        return self._reinforce(id, amount)

    def forget_memory(self, id: str) -> bool:
        """Explicitly forget (delete) a memory.

        Args:
            id: Memory ID.

        Returns:
            True if memory existed and was deleted.
        """
        if id in self._memory_state:
            del self._memory_state[id]
        return self._index.delete(id)

    def get_memory(self, id: str) -> Optional[Dict[str, Any]]:
        """Get a specific memory by ID.

        Returns:
            Memory dict or None.
        """
        info = self._index.get(id)
        if info is None:
            return None

        state = self._memory_state.get(id, {})
        return {
            "id": id,
            "metadata": info["metadata"],
            "importance": state.get("importance", 0.0),
            "access_count": state.get("access_count", 0),
            "last_accessed": state.get("last_accessed"),
            "created_at": state.get("created_at"),
        }

    def get_stats(self) -> Dict[str, Any]:
        """Get memory system statistics."""
        if not self._memory_state:
            return {"total": 0, "capacity": self.capacity}

        importances = [s["importance"] for s in self._memory_state.values()]
        return {
            "total": len(self._memory_state),
            "capacity": self.capacity,
            "mean_importance": float(np.mean(importances)),
            "min_importance": float(np.min(importances)),
            "max_importance": float(np.max(importances)),
            "total_accesses": sum(s["access_count"] for s in self._memory_state.values()),
        }

    def _reinforce(self, id: str, amount: Optional[float] = None) -> bool:
        if id not in self._memory_state:
            return False

        state = self._memory_state[id]
        boost = amount if amount is not None else self.reinforce_amount
        state["importance"] = min(1.0, state["importance"] + boost)
        state["access_count"] += 1
        state["last_accessed"] = time.time()
        return True

    def _apply_decay(self) -> None:
        """Decay all memory importances based on time since last access."""
        now = time.time()
        for state in self._memory_state.values():
            days_since_access = (now - state["last_accessed"]) / 86400.0
            decay = self.decay_rate * days_since_access
            state["importance"] = max(0.0, state["importance"] - decay)

    def _evict(self) -> None:
        """Evict lowest-importance memories until under capacity."""
        while len(self._index) > self.capacity:
            # Find lowest importance memory
            worst_id = None
            worst_importance = float("inf")
            for mid, state in self._memory_state.items():
                if state["importance"] < worst_importance:
                    worst_importance = state["importance"]
                    worst_id = mid

            if worst_id is None:
                break

            self.forget_memory(worst_id)

    def __len__(self) -> int:
        return len(self._index)

File: src/memory.rs
//! Agent memory with prioritization, aging, and eviction.
//!
//! Memories have importance scores that decay over time. Frequently
//! accessed memories get reinforced. Low-value memories are evicted
//! when capacity is reached.

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::streaming::StreamingIndex;

/// State for a single memory entry.
#[derive(Clone, Debug)]
struct MemoryState {
    importance: f64,
    access_count: u64,
    last_accessed: f64,
    created_at: f64,
}

/// A retrieved memory result.
#[derive(Clone, Debug)]
pub struct MemoryResult {
    pub id: String,
    pub content: String,
    pub importance: f64,
    pub score: f32,
    pub access_count: u64,
    pub metadata: HashMap<String, String>,
}

/// Memory system statistics.
#[derive(Clone, Debug)]
pub struct MemoryStats {
    pub total: usize,
    pub capacity: usize,
    pub mean_importance: f64,
    pub min_importance: f64,
    pub max_importance: f64,
    pub total_accesses: u64,
}

/// Prioritized agent memory with decay and eviction.
pub struct AgentMemory {
    dim: usize,
    capacity: usize,
    decay_rate: f64,
    reinforce_amount: f64,
    index: StreamingIndex,
    states: HashMap<String, MemoryState>,
}

fn now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

impl AgentMemory {
    pub fn new(
        dim: usize,
        capacity: usize,
        decay_rate: f64,
        reinforce_amount: f64,
        rerank_factor: usize,
    ) -> Self {
        Self {
            dim,
            capacity,
            decay_rate,
            reinforce_amount,
            index: StreamingIndex::new(dim, rerank_factor),
            states: HashMap::new(),
        }
    }

    /// Store a new memory.
    pub fn save_memory(
        &mut self,
        vector: &[f32],
        content: &str,
        importance: f64,
        id: Option<String>,
        metadata: Option<HashMap<String, String>>,
    ) -> String {
        let importance = importance.clamp(0.0, 1.0);
        let ts = now();

        let mut meta = metadata.unwrap_or_default();
        meta.insert("content".to_string(), content.to_string());

        let mid = self.index.insert(vector, id, Some(meta));

        self.states.insert(mid.clone(), MemoryState {
            importance,
            access_count: 0,
            last_accessed: ts,
            created_at: ts,
        });

        // Evict if over capacity
        while self.index.len() > self.capacity {
            self.evict_lowest();
        }

        mid
    }

    /// Retrieve relevant memories and reinforce them.
    pub fn retrieve_memory(
        &mut self,
        query: &[f32],
        k: usize,
        min_importance: f64,
    ) -> Vec<MemoryResult> {
        self.apply_decay();

        let (scores, ids, metas) = self.index.search(query, k * 3);

        // Collect candidates first, then reinforce (avoids borrow conflict)
        let mut candidates: Vec<(f32, String, HashMap<String, String>)> = Vec::new();
        for ((score, id), meta) in scores.into_iter().zip(ids.into_iter()).zip(metas.into_iter()) {
            let state = match self.states.get(&id) {
                Some(s) => s,
                None => continue,
            };

            if state.importance < min_importance {
                continue;
            }

            candidates.push((score, id, meta));
            if candidates.len() >= k {
                break;
            }
        }

        // Now reinforce and build results
        let mut results = Vec::new();
        for (score, id, meta) in candidates {
            self.reinforce(&id);

            let state = self.states.get(&id).unwrap();
            let content = meta.get("content").cloned().unwrap_or_default();
            let filtered_meta: HashMap<String, String> = meta.into_iter()
                .filter(|(k, _)| k != "content")
                .collect();

            results.push(MemoryResult {
                id,
                content,
                importance: state.importance,
                score,
                access_count: state.access_count,
                metadata: filtered_meta,
            });
        }

        results
    }

    /// Manually reinforce a memory's importance.
    pub fn reinforce_memory(&mut self, id: &str, amount: Option<f64>) -> bool {
        self.reinforce_with_amount(id, amount)
    }

    /// Explicitly forget (delete) a memory.
    pub fn forget_memory(&mut self, id: &str) -> bool {
        self.states.remove(id);
        self.index.delete(id)
    }

    /// Get memory system statistics.
    pub fn stats(&self) -> MemoryStats {
        if self.states.is_empty() {
            return MemoryStats {
                total: 0,
                capacity: self.capacity,
                mean_importance: 0.0,
                min_importance: 0.0,
                max_importance: 0.0,
                total_accesses: 0,
            };
        }

        let importances: Vec<f64> = self.states.values().map(|s| s.importance).collect();
        let total_accesses: u64 = self.states.values().map(|s| s.access_count).sum();

        MemoryStats {
            total: self.states.len(),
            capacity: self.capacity,
            mean_importance: importances.iter().sum::<f64>() / importances.len() as f64,
            min_importance: importances.iter().cloned().fold(f64::MAX, f64::min),
            max_importance: importances.iter().cloned().fold(f64::MIN, f64::max),
            total_accesses,
        }
    }

    pub fn len(&self) -> usize {
        self.index.len()
    }

    pub fn is_empty(&self) -> bool {
        self.index.is_empty()
    }

    fn reinforce(&mut self, id: &str) {
        self.reinforce_with_amount(id, None);
    }

    fn reinforce_with_amount(&mut self, id: &str, amount: Option<f64>) -> bool {
        let state = match self.states.get_mut(id) {
            Some(s) => s,
            None => return false,
        };
        let boost = amount.unwrap_or(self.reinforce_amount);
        state.importance = (state.importance + boost).min(1.0);
        state.access_count += 1;
        state.last_accessed = now();
        true
    }

    fn apply_decay(&mut self) {
        let ts = now();
        for state in self.states.values_mut() {
            let days_since_access = (ts - state.last_accessed) / 86400.0;
            let decay = self.decay_rate * days_since_access;
            state.importance = (state.importance - decay).max(0.0);
        }
    }

    fn evict_lowest(&mut self) {
        let worst_id = self.states.iter()
            .min_by(|a, b| a.1.importance.partial_cmp(&b.1.importance).unwrap())
            .map(|(id, _)| id.clone());

        if let Some(id) = worst_id {
            self.forget_memory(&id);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_save_and_retrieve() {
        let dim = 8;
        let mut mem = AgentMemory::new(dim, 100, 0.05, 0.1, 10);

        let v = vec![1.0; 8];
        let id = mem.save_memory(&v, "test memory", 0.8, None, None);
        assert_eq!(mem.len(), 1);

        let results = mem.retrieve_memory(&v, 1, 0.0);
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].content, "test memory");
    }

    #[test]
    fn test_eviction() {
        let dim = 8;
        let mut mem = AgentMemory::new(dim, 2, 0.0, 0.1, 10);

        let v1 = vec![1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];
        let v2 = vec![0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];
        let v3 = vec![0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0];

        mem.save_memory(&v1, "low", 0.1, None, None);
        mem.save_memory(&v2, "high", 0.9, None, None);
        mem.save_memory(&v3, "medium", 0.5, None, None);

        // Capacity is 2, so lowest importance should be evicted
        assert_eq!(mem.len(), 2);
    }
}
